lib/logstash/inputs/udp.rb

```diff
@@ -81,26 +81,28 @@ def stop
   def udp_listener(output_queue)
     @logger.info("Starting UDP listener", :address => "#{@host}:#{@port}")
 
-    if @udp && !@udp.closed?
+    if @udp && !@udp.isClosed
       @udp.close
     end
 
-    if IPAddr.new(@host).ipv6?
-      @udp = UDPSocket.new(Socket::AF_INET6)
-    elsif IPAddr.new(@host).ipv4?
-      @udp = UDPSocket.new(Socket::AF_INET)
-    end
+    # Bind to nil to defer binding until later
+    @udp = java.net.DatagramSocket.new(nil)
+
     # set socket receive buffer size if configured
     if @receive_buffer_bytes
-      @udp.setsockopt(Socket::SOL_SOCKET, Socket::SO_RCVBUF, @receive_buffer_bytes)
-    end
-    rcvbuf = @udp.getsockopt(Socket::SOL_SOCKET, Socket::SO_RCVBUF).unpack("i")[0]
-    if @receive_buffer_bytes && rcvbuf != @receive_buffer_bytes
-      @logger.warn("Unable to set receive_buffer_bytes to desired size. Requested #{@receive_buffer_bytes} but obtained #{rcvbuf} bytes.")
+      @udp.setReceiveBufferSize(@receive_buffer_bytes)
+
+      actual_receive_buffer_bytes = @udp.getReceiveBufferBytes
+      if  actual_receive_buffer_bytes != @receive_buffer_bytes
+        @logger.warn("Unable to set receive_buffer_bytes to desired size. Requested #{@receive_buffer_bytes} but obtained #{actual_receive_buffer_bytes} bytes.")
+      end
     end
+    
+    @udp.bind(java.net.InetSocketAddress.new(@host, @port))
+
+    @udp.setSoTimeout(10) # only block for 10ms per read
 
-    @udp.bind(@host, @port)
-    @logger.info("UDP listener started", :address => "#{@host}:#{@port}", :receive_buffer_bytes => "#{rcvbuf}", :queue_size => "#{@queue_size}")
+    @logger.info("UDP listener started", :address => "#{@host}:#{@port}", :receive_buffer_bytes => "#{@receive_buffer_bytes}", :queue_size => "#{@queue_size}")
```

Comment:
To keep to previous behavior we should be log `actual_receive_buffer_bytes` not `@receive_buffer_bytes`